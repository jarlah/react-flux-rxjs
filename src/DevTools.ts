export type Message = {
  type: string
  payload: {
    type: string
  }
  state: any
}

export type DevToolsInstance = {
  subscribe: (sub: (message: Message) => void) => () => void
  send: (n: string, o: any) => void
}

export type DevToolsExtension = {
  connect: (config?: { name?: string }) => DevToolsInstance
  disconnect: () => void
}

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION__?: DevToolsExtension
    devToolsExtension?: DevToolsExtension
  }
}

export function isRelevant(message: Message): boolean {
  if (message.type === "DISPATCH") {
    switch (message.payload.type) {
      case "JUMP_TO_ACTION":
      case "JUMP_TO_STATE":
        return true
      default:
        return false
    }
  }
  return false
}

export function getExtension(): DevToolsExtension | null {
  if (process.env.NODE_ENV === "development" && typeof window !== "undefined") {
    const ext = window.__REDUX_DEVTOOLS_EXTENSION__ || window.devToolsExtension
    if (ext) {
      return ext
    }
  }
  return null
}
