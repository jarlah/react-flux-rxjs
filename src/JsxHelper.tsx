import * as React from "react"

export type CT<P> = React.ComponentType<P>

export function render<P>(C: CT<P>, p: P): React.ReactElement<P> {
  return <C {...p} />
}
