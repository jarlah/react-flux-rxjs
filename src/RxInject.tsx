import * as React from "react"
import { Observable, Subscription } from "rxjs"

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION__?: DevTools | undefined
    devToolsExtension?: DevTools | undefined
  }
}

export type Injector<ComponentProps, ParentProps> = (
  Component: React.ComponentClass<ComponentProps>
) => React.ComponentClass<ParentProps>

export type PropsType<ComponentProps, StoreProps, UpstreamProps> = (
  store: StoreProps,
  upstream: UpstreamProps
) => ComponentProps

export type Message = {
  type: string
  payload: {
    type: string
  }
  state: any
}

export type DevToolsInstance = {
  subscribe: (sub: (message: Message) => void) => () => void
  send: (n: string, o: any) => void
}

export type DevTools = {
  connect: () => DevToolsInstance
  disconnect: () => void
}

export type Store<T> = Observable<T> | StoreFactory<T>
export type StoreFactory<T> = () => Observable<T>

export default function inject<ComponentProps, StoreProps, ParentProps>(
  store: Store<StoreProps>,
  props: PropsType<ComponentProps, StoreProps, ParentProps>
): Injector<ComponentProps, ParentProps> {
  return (Component: React.ComponentClass<ComponentProps>) => {
    type State = { store: StoreProps }
    class Inject extends React.Component<ParentProps, State> {
      state: State
      subscription: Subscription
      unsubscribe: () => void
      devTools: DevToolsInstance

      componentWillMount() {
        const devToolsExt = getDevToolsExt()
        if (devToolsExt) {
          this.devTools = devToolsExt.connect()
          this.unsubscribe = this.devTools.subscribe(message => {
            if (
              message.type === "DISPATCH" &&
              (message.payload.type === "JUMP_TO_ACTION" ||
                message.payload.type === "JUMP_TO_STATE")
            ) {
              const props: StoreProps = JSON.parse(message.state)
              this.setState({ store: props })
            }
          })
        }
      }

      componentDidMount() {
        const observable = getObservable(store)
        this.subscription = observable.subscribe(storeProps => {
          if (this.devTools) {
            this.devTools.send("update", storeProps)
          }
          this.setState({ store: storeProps })
        })
      }

      componentWillUnmount() {
        this.subscription.unsubscribe()
        const devToolsExt = getDevToolsExt()
        if (devToolsExt) {
          this.unsubscribe()
          devToolsExt.disconnect()
        }
      }

      render() {
        if (!this.state) {
          return null
        }
        const customProps =
          typeof props === "function"
            ? props(this.state.store, this.props)
            : props
        return <Component {...customProps} />
      }
    }
    return Inject
  }
}

function getDevToolsExt(): DevTools | undefined {
  if (process.env.NODE_ENV === "development" && typeof window !== "undefined") {
    return window.__REDUX_DEVTOOLS_EXTENSION__ && window.devToolsExtension
  }
}

function getObservable<T>(store: Store<T>): Observable<T> {
  let observable: Observable<T>
  if (store instanceof Observable) {
    observable = store
  } else if (typeof store === "function") {
    observable = store()
  } else {
    observable = store
  }
  return observable
}
