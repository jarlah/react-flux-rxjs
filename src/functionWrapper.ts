import * as React from "react"

function getDisplayName<P>(Component: React.StatelessComponent<P>): string {
  return Component.displayName || Component.name || "Unknown"
}

const functionWrapper = function<P>(
  functionalComponent: React.StatelessComponent<P>
): React.ComponentClass<P> {
  return class extends React.Component<P, {}> {
    static displayName = getDisplayName(functionalComponent)
    render() {
      return functionalComponent(this.props)
    }
  }
}

export default functionWrapper
