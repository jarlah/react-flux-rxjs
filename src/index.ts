import * as React from 'react';
import { useCallback, useEffect, useState } from 'react';
import { Observable } from 'rxjs';
import { publishReplay, refCount, scan, startWith, tap } from 'rxjs/operators';

export type Injector<ComponentProps, ParentProps> = (
  Component: React.ComponentType<ComponentProps>,
) => React.ComponentType<ParentProps>;

export type PropsFactory<ComponentProps, StoreProps, UpstreamProps> = (
  store: StoreProps,
  upstream: UpstreamProps,
) => ComponentProps;

export type PropsType<ComponentProps, StoreProps, UpstreamProps> =
  | PropsFactory<ComponentProps, StoreProps, UpstreamProps>
  | ComponentProps;

export type Store<ParentProps, StoreProps> = Observable<StoreProps> | StoreFactory<ParentProps, StoreProps>;

export type StoreFactory<ParentProps, StoreProps> = (props: ParentProps) => Observable<StoreProps>;

export type Reducer<T> = (state: T) => T;

export function inject<ComponentProps, StoreProps, ParentProps>(
  store: Store<ParentProps, StoreProps>,
  props: PropsType<ComponentProps, StoreProps, ParentProps>,
): Injector<ComponentProps, ParentProps> {
  return (Component: React.ComponentType<ComponentProps>) => {
    return (parentProps: ParentProps) => {
      const [state, setState] = useState<{ store: StoreProps }>();

      const updateState = useCallback((storeProps: StoreProps) => {
        setState({ store: storeProps });
      }, []);

      useEffect(() => {
        const observable = typeof store !== 'function' ? store : store(parentProps);
        const storeSubscription = observable.pipe(tap(updateState)).subscribe();
        return () => storeSubscription.unsubscribe();
      }, []);

      if (!state) {
        return null;
      }

      const customProps =
        typeof props === 'function'
          ? (props as PropsFactory<ComponentProps, StoreProps, ParentProps>)(state.store, parentProps)
          : props;

      return React.createElement(Component, customProps, null);
    };
  };
}

export function createStore<T extends any>(
  name: string,
  reducer$: Observable<Reducer<T>>,
  initialState?: T,
  keepAlive: boolean = false,
): Observable<T> {
  initialState = typeof initialState !== 'undefined' ? initialState : ({} as T);
  const store = reducer$.pipe(
    scan((state: T, reducer: Reducer<T>) => reducer(state), initialState),
    startWith(initialState),
    publishReplay(1),
    refCount(),
  );
  if (keepAlive) {
    store.subscribe();
  }
  return store;
}
