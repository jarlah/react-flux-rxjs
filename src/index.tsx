import * as React from 'react';
import { useEffect, useState } from 'react';
import { Observable, ReplaySubject } from 'rxjs';
import { scan, share, startWith, tap } from 'rxjs/operators';

export type Injector<ComponentProps, ParentProps> = (
  Component: React.ComponentType<ComponentProps>,
) => React.ComponentType<ParentProps>;

export type PropsFactory<ComponentProps, StoreProps, UpstreamProps> = (
  store: StoreProps,
  upstream: UpstreamProps,
) => ComponentProps;

export type PropsType<ComponentProps, StoreProps, UpstreamProps> =
  | PropsFactory<ComponentProps, StoreProps, UpstreamProps>
  | ComponentProps;

export type Store<ParentProps, StoreProps> = Observable<StoreProps> | StoreFactory<ParentProps, StoreProps>;

export type StoreFactory<ParentProps, StoreProps> = (props: ParentProps) => Observable<StoreProps>;

export type Reducer<T> = (state: T) => T;

/**
 * this method should work similar to connect in react-redux, albeit a bit differently.
 *
 * @param store the store
 * @param props the props
 */
export function connect<ComponentProps, StoreProps, ParentProps>(
  store: Store<ParentProps, StoreProps>,
  props: PropsType<ComponentProps, StoreProps, ParentProps>,
): Injector<ComponentProps, ParentProps> {
  return (Component: React.ComponentType<ComponentProps>) => {
    return (parentProps: ParentProps) => {
      const [state, setState] = useState<StoreProps>();

      useEffect(() => {
        const observable = typeof store !== 'function' ? store : store(parentProps);
        const storeSubscription = observable.pipe(tap(setState)).subscribe();
        return () => storeSubscription.unsubscribe();
      }, []);

      if (!state) {
        // wait for initial data before render the component
        return null;
      }

      const customProps =
        typeof props === 'function'
          ? (props as PropsFactory<ComponentProps, StoreProps, ParentProps>)(state, parentProps)
          : props;

      return <Component {...customProps} />;
    };
  };
}

/**
 * This method basically creates an observable,
 * but it does a lot of footwork to simulate how a store (in lets say redux) works.
 *
 * @param name the name of the store
 * @param reducer$ the observable reducer
 * @param initialState the initial state, can be plain object or any type
 * @param keepAlive if this store is reused across multiple mount/unmount, should this store remember its state?
 */
export function createStore<T>(
  name: string,
  reducer$: Observable<Reducer<T>>,
  initialState: T = null as any,
  keepAlive: boolean = false,
): Observable<T> {
  const store = reducer$.pipe(
    scan((state: T, reducer: Reducer<T>) => reducer(state), initialState),
    startWith(initialState),
    share({
      connector: () => new ReplaySubject(1),
      resetOnError: false,
      resetOnComplete: false,
      resetOnRefCountZero: false,
    })
  );
  if (keepAlive) {
    store.subscribe();
  }
  return store;
}
