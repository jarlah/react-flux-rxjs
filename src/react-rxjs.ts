import * as React from "react"
import { Observable, Subscription } from "rxjs"
import { publishReplay, refCount, scan, startWith, tap } from "rxjs/operators"

export type Injector<ComponentProps, ParentProps> = (
  Component: React.ComponentType<ComponentProps>
) => React.ComponentType<ParentProps>

export type PropsFactory<ComponentProps, StoreProps, UpstreamProps> = (
  store: StoreProps,
  upstream: UpstreamProps
) => ComponentProps

export type PropsType<ComponentProps, StoreProps, UpstreamProps> =
  | PropsFactory<ComponentProps, StoreProps, UpstreamProps>
  | ComponentProps

export type Store<ParentProps, StoreProps> =
  | Observable<StoreProps>
  | StoreFactory<ParentProps, StoreProps>

export type StoreFactory<ParentProps, StoreProps> = (props: ParentProps) => Observable<StoreProps>

export type Reducer<T> = (state: T) => T

export function inject<ComponentProps, StoreProps, ParentProps>(
  store: Store<ParentProps, StoreProps>,
  props: PropsType<ComponentProps, StoreProps, ParentProps>
): Injector<ComponentProps, ParentProps> {
  return (Component: React.ComponentType<ComponentProps>) => {
    class Inject extends React.Component<ParentProps, { store: StoreProps }> {
      storeSubscription?: Subscription
      devToolsSubscription?: () => void

      updateState = (store: StoreProps) => {
        this.setState({ store })
      }

      componentDidMount() {
        const observable = typeof store !== "function" ? store : store(this.props)
        this.storeSubscription = observable.pipe(tap(this.updateState)).subscribe()
      }

      componentWillUnmount() {
        // @ts-ignore if this is null, then we haven't passed componentDidMount and we are not here
        this.storeSubscription.unsubscribe()
      }

      render() {
        if (!this.state) {
          return null
        }
        const customProps =
          typeof props === "function"
            ? (props as PropsFactory<ComponentProps, StoreProps, ParentProps>)(
                this.state.store,
                this.props
              )
            : props
        return React.createElement(Component, customProps, null)
      }
    }
    return Inject
  }
}

export function createStore<T extends Object>(
  name: string,
  reducer$: Observable<Reducer<T>>,
  initialState?: T,
  keepAlive: boolean = false
): Observable<T> {
  initialState = typeof initialState !== "undefined" ? initialState : ({} as T)
  const store = reducer$.pipe(
    scan((state: T, reducer: Reducer<T>) => reducer(state), initialState),
    startWith(initialState),
    publishReplay(1),
    refCount()
  )
  if (keepAlive) {
    store.subscribe()
  }
  return store
}
