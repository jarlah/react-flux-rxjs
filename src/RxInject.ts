import * as React from "react"
import { render, getName } from "./JsxHelper"
import { Observable, Subscription } from "rxjs"
import { Store, PropsType, Injector } from "./RxTypes"
import {
  DevToolsExtension,
  DevToolsInstance,
  getExtension,
  isRelevant
} from "./DevTools"

export default function inject<ComponentProps, StoreProps, ParentProps>(
  store: Store<ParentProps, StoreProps>,
  props: PropsType<ComponentProps, StoreProps, ParentProps>,
  _devTools?: DevToolsExtension
): Injector<ComponentProps, ParentProps> {
  const devTools: DevToolsExtension | null = _devTools || getExtension()
  return (Component: React.ComponentType<ComponentProps>) => {
    class Inject extends React.Component<ParentProps, { store: StoreProps }> {
      storeSubscription: Subscription
      devToolsSubscription: () => void
      devToolsInstance: DevToolsInstance

      componentWillMount() {
        if (devTools) {
          this.devToolsInstance = devTools.connect({
            name: `${getName(Component)}Container`
          })
          this.devToolsSubscription = this.devToolsInstance.subscribe(
            message => {
              if (isRelevant(message)) {
                const props: StoreProps = JSON.parse(message.state)
                this.setState({ store: props })
              }
            }
          )
        }
      }

      sendToDevTools(store: StoreProps) {
        this.devToolsInstance && this.devToolsInstance.send("update", store)
      }

      updateState(store: StoreProps) {
        this.setState({ store })
      }

      componentDidMount() {
        const observable = getObservable(store, this.props)
        this.storeSubscription = observable
          .do(this.sendToDevTools.bind(this))
          .do(this.updateState.bind(this))
          .subscribe()
      }

      componentWillUnmount() {
        this.storeSubscription.unsubscribe()
        if (devTools) {
          this.devToolsSubscription()
          devTools.disconnect()
        }
      }

      render() {
        if (!this.state) {
          return null
        }
        const customProps =
          typeof props === "function"
            ? props(this.state.store, this.props)
            : props
        return render(Component, customProps)
      }
    }
    return Inject
  }
}

function getObservable<P, T>(store: Store<P, T>, parentPops: P): Observable<T> {
  return store instanceof Observable ? store : store(parentPops)
}
