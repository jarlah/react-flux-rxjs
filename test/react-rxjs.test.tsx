import { inject, createStore } from "../src/react-rxjs"
import {EMPTY, of, Subject} from "rxjs"
import * as React from "react"
import {configure, mount} from "enzyme"
import shallowToJson from "enzyme-to-json"
import {map} from "rxjs/operators";
import * as Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

beforeEach(() => {
  process.env.NODE_ENV = "development"
});

afterEach(() => {
  delete process.env.NODE_ENV
});

describe("RxInject", () => {
  it("is instantiable with Observable", done => {
    let wrapper;

    const NumberComp = (props: { number: number }) => (
      <span>{props.number}</span>
    );

    const stream = of(0, 1, 2);

    const InjectedNumberComp = inject(
      stream,
      (storeProps: number) => ({
        number: storeProps
      })
    )(NumberComp);

    expect(InjectedNumberComp).toBeInstanceOf(Function);

    wrapper = mount(<InjectedNumberComp />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    setTimeout(() => {
      wrapper.unmount();
      done()
    }, 500)
  })

  it("is instantiable with props object", () => {
    const NumberComp = (props: { count: number }) => (
      <span>{props.count}</span>
    );

    const InjectedNumberComp = inject(of({ count: 0 } ), { count: 5000 })(
      NumberComp
    );

    expect(InjectedNumberComp).toBeInstanceOf(Function);

    const wrapper = mount(<InjectedNumberComp />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    wrapper.unmount()
  });

  it("is instantiable with props object and nulled devtools", () => {
    const NumberComp = (props: { number: number }) => (
      <span>{props.number}</span>
    );

    const InjectedNumberComp = inject(of(0), { number: 5000 })(
      NumberComp
    );

    expect(InjectedNumberComp).toBeInstanceOf(Function);

    const wrapper = mount(<InjectedNumberComp />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    wrapper.unmount()
  });

  it("is instantiable with Factory abd props function", () => {
    const NumberComp = (props: { number: number }) => (
      <span>{props.number}</span>
    );

    const InjectedNumberComp = inject(
      () => of(666),
      (storeProps: number) => ({ number: storeProps })
    )(NumberComp);

    expect(InjectedNumberComp).toBeInstanceOf(Function);

    const wrapper = mount(<InjectedNumberComp />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    wrapper.unmount()
  });

  it("is instantiable with Factory and class component", () => {
    class NumberComp2 extends React.Component<{ number: number }, {}> {
      render() {
        return <span>{this.props.number}</span>
      }
    }

    const InjectedNumberComp = inject(
      () => of(1337),
      (storeProps: number) => ({ number: storeProps })
    )(NumberComp2);

    expect(InjectedNumberComp).toBeInstanceOf(Function);

    const wrapper = mount(<InjectedNumberComp />);
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
});

describe("RxStore", () => {
  it("has initial state", done => {
    const store$ = createStore("test", EMPTY, 42)

    store$.subscribe((n: number) => {
      expect(n).toBe(42);
      done()
    })
  });

  it("gets state updates from reducer", done => {
    let next = 0;

    const action = new Subject<void>();

    const store$ = createStore(
      "test",
      action.pipe(map(() => (state: number) => state + 1)),
      next,
      true
    );

    store$.subscribe((n: number) => {
      expect(n).toBe(next);
      if (n === 2) {
        done()
      } else {
        next++
      }
    });
    setTimeout(action.next.bind(action), 0);
    setTimeout(action.next.bind(action), 0)
  })
});
