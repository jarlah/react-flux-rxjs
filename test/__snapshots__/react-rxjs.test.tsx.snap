// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RxInject is instantiable with Factory abd props function 1`] = `
<Inject>
  <NumberComp
    number={666}
  >
    <span>
      666
    </span>
  </NumberComp>
</Inject>
`;

exports[`RxInject is instantiable with Factory and class component 1`] = `
<Inject>
  <NumberComp2
    number={1337}
  >
    <span>
      1337
    </span>
  </NumberComp2>
</Inject>
`;

exports[`RxInject is instantiable with Observable 1`] = `
<Inject>
  <NumberComp
    number={2}
  >
    <span>
      2
    </span>
  </NumberComp>
</Inject>
`;

exports[`RxInject is instantiable with props object 1`] = `
<Inject>
  <NumberComp
    count={5000}
  >
    <span>
      5000
    </span>
  </NumberComp>
</Inject>
`;

exports[`RxInject is instantiable with props object and nulled devtools 1`] = `
<Inject>
  <NumberComp
    number={5000}
  >
    <span>
      5000
    </span>
  </NumberComp>
</Inject>
`;
